src/test/java/com/xwiki/authentication/saml/XWikiMock.java

```diff
@@ -0,0 +1,132 @@
+package com.xwiki.authentication.saml;
+
+import com.xpn.xwiki.XWiki;
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.objects.BaseObject;
+import com.xpn.xwiki.objects.StringProperty;
+import com.xpn.xwiki.objects.classes.BaseClass;
+import com.xpn.xwiki.objects.classes.StringClass;
+import com.xpn.xwiki.objects.meta.StringMetaClass;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.EntityReference;
+import org.xwiki.rendering.syntax.Syntax;
+
+import java.util.*;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+
+public class XWikiMock extends XWiki {
+    final Map<String, List<BaseObject>> createdObjectsByEntity = new LinkedHashMap<>();
+    public Set<XWikiDocument> savedDocuments = new HashSet<>();
+    private final BaseClass newClass;
+    private final Map<DocumentReference, XWikiDocument> docByReference = new HashMap<>();
+
+    final BaseObject baseObjectMock = mock(BaseObject.class);
+    private final Map<String, String> savedUserAttributes = new LinkedHashMap<>();
+
+    public XWikiMock(XWikiContext context) {
+        newClass = new BaseClass() {
+            public DocumentReference getDocumentReference() {
+                return new DocumentReference(context.getWikiId(), "XWiki:", "XWikiGroups.StarshipTroopers");
+            }
+
+            @SuppressWarnings("rawtypes")
+            @Override
+            public Collection getFieldList() {
+                Map<String, Object> fields = new LinkedHashMap<>();
+
+                StringMetaClass metaClass = new StringMetaClass();
+                StringClass prop = new StringClass("member", "XWikiGroups", metaClass);
+                fields.put("member", prop);
+
+                return fields.values();
+            }
+        };
+
+        doAnswer(invocation -> {
+            savedUserAttributes.put(invocation.getArguments()[0].toString(), invocation.getArguments()[1].toString());
+            return null;
+        }).when(baseObjectMock).set(any(), any(), any());
+    }
+
+    @Override
+    public int createUser(String userName,
+                          Map<String, ?> map,
+                          EntityReference parentReference,
+                          String content,
+                          Syntax syntax,
+                          String userRights,
+                          XWikiContext context) {
+        map.forEach((key, value) -> this.savedUserAttributes.put(key + "", value + ""));
```

Comment:
This isn't Javascript (and even then I'd dislike this), don't use add operators to convert to string, it isn't obvious what you are doing.